libraries/AP_Math/examples/throttle_curve/throttle_curve.cpp

```diff
@@ -0,0 +1,108 @@
+//
+// Unit tests for the AP_Math throttle curve code
+//
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Math/AP_Math.h>
+
+void test_throttle_curve(float alpha);
+void setup();
+void loop();
+
+const AP_HAL::HAL& hal = AP_HAL::get_HAL();
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+
+#include <iostream>
+#include <fstream>
+#include <vector>
+using namespace std;
+
+void print_line(std::vector<double> &vec, ofstream &f);
+float copter_throttle_curve(float thr_mid, float throttle_in);
+
+// This code must be copied from Copter's Mode::get_pilot_desired_throttle() since
+// there is no method analogous to AP_Math:throttle_curve in copter
+float copter_throttle_curve(float thr_mid, float throttle_in)
+{
+    const float expo = constrain_float(-(thr_mid-0.5f)/0.375f, -0.5f, 1.0f);
+    // calculate the output throttle using the given expo function
+    float throttle_out = throttle_in*(1.0f-expo) + expo*throttle_in*throttle_in*throttle_in;
+    return throttle_out;
+}
+
+void print_line(vector<double> &vec, ofstream &f)
+{
+    for (double v : vec) {
+        f << v << ",";
+    }
+    f << "\n";
+}
+
+void test_throttle_curve(float alpha)
+{
+    ofstream csv_out;
+
+    const int ns = 101;
```

Comment:
normally we don't use "int".. more often it's int8_t or int16_t..

Reply:
I wasn't intending for this code to be cross-compiled for a flight controller, and used int for convenience.  But I hadn't given any thought to the implications of using signed/unsigned byte/word/longword integers as loop indexes.  I wonder if there's a penalty for using a non-native type?

I guess the vectors of vectors of double would be pretty bad on a flight controller too, so maybe I should warn people that this is not an example of how to write flight code?

Reply:
Warning is definitely in order.  And removing the structure infered by running on a flight controller will simplify this.

In terms of typing for loop counters - the pattern we use throughout the codebase is uint8_t - right or wrong - so we should stick to that.

MdB looked at whether it made a difference in the code at one stage - it didn't AFAIR.